clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -400,12 +412,33 @@ private void handleFetchResponse(ClientResponse resp, FetchRequest request) {
             int totalBytes = 0;
             int totalCount = 0;
             FetchResponse response = new FetchResponse(resp.responseBody());
+
+            // Look for OffsetOutOfRange error in fetchResponse
+            Map<TopicPartition, Long> offsetOutOfRangePartitions = new HashMap();
+            for (Map.Entry<TopicPartition, FetchResponse.PartitionData> entry : response.responseData().entrySet()) {
+                TopicPartition tp = entry.getKey();
+                FetchResponse.PartitionData partition = entry.getValue();
+                long fetchOffset = request.fetchData().get(tp).offset;
+                if (subscriptions.assignedPartitions().contains(tp) && partition.errorCode == Errors.OFFSET_OUT_OF_RANGE.code())
+                    offsetOutOfRangePartitions.put(tp, fetchOffset);
+            }
+            if (!offsetOutOfRangePartitions.isEmpty())
+                this.exceptionFromLastFetch = new OffsetOutOfRangeException(offsetOutOfRangePartitions);
```

Comment:
Maybe instead of storing the exception as a field, you can store the map and create the exception in fetched records if the map is empty?